### PR TITLE
perf(blooms): Avoid tiny string allocations for insert cache

### PR DESCRIPTION
**Pull request by chaudum**
**What this PR does / why we need it**:

Given that we see really high insert cache hit rate, we can void the string allocations for the cache ownership check.

screenshot_20240711_135526

```
sum(rate(loki_bloom_inserts_total{collision="cache"}[$__rate_interval]))
/
sum(rate(loki_bloom_inserts_total{}[$__rate_interval]))
```

**Special notes for your reviewer**:

Running benchmark with command:

```console
$ gotest -v -count=10 ./pkg/storage/bloom/v1/... -test.run=^$ -test.bench=BenchmarkPopulateSeriesWithBloom
```

Results:

```console
$ benchstat bench-old.txt bench-new.txt 
bench-old.txt:5: missing iteration count
bench-new.txt:5: missing iteration count
goos: linux
goarch: amd64
pkg: github.com/grafana/loki/v3/pkg/storage/bloom/v1
cpu: 11th Gen Intel(R) Core(TM) i7-1185G7 @ 3.00GHz
                          │ bench-old.txt │           bench-new.txt            │
                          │    sec/op     │   sec/op     vs base               │
PopulateSeriesWithBloom-8   1038.9µ ± 29%   973.7µ ± 2%  -6.28% (p=0.000 n=10)

                          │ bench-old.txt │            bench-new.txt            │
                          │     B/op      │     B/op      vs base               │
PopulateSeriesWithBloom-8    9.112Mi ± 0%   9.054Mi ± 0%  -0.64% (p=0.000 n=10)

                          │ bench-old.txt │           bench-new.txt            │
                          │   allocs/op   │ allocs/op   vs base                │
PopulateSeriesWithBloom-8     5269.0 ± 0%   877.0 ± 0%  -83.36% (p=0.000 n=10)
```

**Checklist**
- [ ] Reviewed the [`CONTRIBUTING.md`](https://github.com/grafana/loki/blob/main/CONTRIBUTING.md) guide (**required**)
- [ ] Documentation added
- [ ] Tests updated
- [x] Title matches the required conventional commits format, see [here](https://www.conventionalcommits.org/en/v1.0.0/)
  - **Note** that Promtail is considered to be feature complete, and future development for logs collection will be in [Grafana Alloy](https://github.com/grafana/alloy). As such, `feat` PRs are unlikely to be accepted unless a case can be made for the feature actually being a bug fix to existing behavior.
- [ ] Changes that require user attention or interaction to upgrade are documented in `docs/sources/setup/upgrade/_index.md`
- [ ] For Helm chart changes bump the Helm chart version in `production/helm/loki/Chart.yaml` and update `production/helm/loki/CHANGELOG.md` and `production/helm/loki/README.md`. [Example PR](https://github.com/grafana/loki/commit/d10549e3ece02120974929894ee333d07755d213)
- [ ] If the change is deprecating or removing a configuration option, update the `deprecated-config.yaml` and `deleted-config.yaml` files respectively in the `tools/deprecated-config-checker` directory. [Example PR](https://github.com/grafana/loki/pull/10840/commits/0d4416a4b03739583349934b96f272fb4f685d15)

### CODE REVIEW

**Review by ashwanthgoli — APPROVED** [review]

**Comment by chaudum** [comment]
@owen-d Can you take a look as well? Feel free to merge it thereafter.

File: pkg/storage/bloom/v1/bloom_tokenizer.go [hunk 1]
```diff
@@ -216,10 +217,12 @@ outer:
 			for itr.Next() {
 				tok := itr.At()
 				tokens++
+
 				// TODO[owen-d]: [n]byte this
-				str := string(tok)
-				_, found := bt.cache[str] // A cache is used ahead of the SBF, as it cuts out the costly operations of scaling bloom filters
-				if found {
+				// To avoid allocations, an unsafe string can be used to check ownership in cache.
+				str := unsafe.String(unsafe.SliceData(tok), len(tok))
```
Comment (owen-d) [inline pkg/storage/bloom/v1/bloom_tokenizer.go:223]: note: The `[]byte` is reused across `Next` calls. I wonder if this impacts us here.
```
// The Token iterator uses shared buffers for performance. The []byte returned by At()
// is not safe for use after subsequent calls to Next()
```

Comment (chaudum) [reply pkg/storage/bloom/v1/bloom_tokenizer.go:223]: The unsafe string is only used for the membership check, not the assignment in the map.

Comment (owen-d) [reply pkg/storage/bloom/v1/bloom_tokenizer.go:223]: Oh good point. LGTM

**Review by owen-d — APPROVED** [review]

**Pull request merged**